### PR TITLE
Add offset writes and file size support to NonBlockingFileIO

### PR DESCRIPTION
**Pull request by sveamarcus**
Motivation:

NonBlockingFileIO currently only supports writes from 0 or appending,
if opened with O_APPEND. To facilitate random read/write file access
the API needs to evolve.

Modifications:

- Add toOffset function parameter to write for offset file write access
- Add changeFileSize function for growing or shrinking the file
- Add readFileSize function for asynchronously reading the file size

Result:

- Improve API for random access scenarios
- Provide method for reading and writing file size

### CODE REVIEW

File: Sources/NIO/NonBlockingFileIO.swift [hunk 1]
```diff
@@ -359,6 +400,29 @@ public struct NonBlockingFileIO {
...
+    ///   - fileHandle: The `NIOFileHandle` to write to.
+    ///   - toOffset: The file offset to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    ///   - eventLoop: The `EventLoop` to create the returned `EventLoopFuture` from.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the write was successful or fails on error.
+    public func write(fileHandle: NIOFileHandle,
+                      toOffset: Int64,
+                      buffer: ByteBuffer,
+                      eventLoop: EventLoop) -> EventLoopFuture<()> {
+        write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer, eventLoop: eventLoop)
```
Comment (Lukasa) [inline Sources/NIO/NonBlockingFileIO.swift]: NIO prefers to use explicit `self`.

File: Sources/NIO/NonBlockingFileIO.swift [hunk 2]
```diff
@@ -359,6 +400,29 @@ public struct NonBlockingFileIO {
     public func write(fileHandle: NIOFileHandle,
                       buffer: ByteBuffer,
                       eventLoop: EventLoop) -> EventLoopFuture<()> {
+        write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer, eventLoop: eventLoop)
```
Comment (Lukasa) [inline Sources/NIO/NonBlockingFileIO.swift]: NIO prefers to use explicit `self`.

File: Sources/NIO/System.swift [hunk 3]
```diff
@@ -353,13 +355,27 @@ internal enum Posix {
         }
     }
 
+    @inline(never)
+    public static func ftruncate(descriptor: CInt, size: off_t) throws -> IOResult<CInt> {
+        return try wrapSyscallMayBlock {
```
Comment (Lukasa) [inline Sources/NIO/System.swift]: This should only be `wrapSyscall`, as there is no chance of this returning `EAGAIN`.

File: Tests/NIOTests/NonBlockingFileIOTest.swift [hunk 4]
```diff
@@ -452,6 +452,45 @@ class NonBlockingFileIOTest: XCTestCase {
...
+    func testChangeFileSizeShrink() throws {
+        try withTemporaryFile(content: "0123456789") { (fileHandle, _) -> Void in
+            try self.fileIO.changeFileSize(fileHandle: fileHandle,
+                                           size: 1,
+                                           eventLoop: eventLoop).wait()
+            let fileRegion = try FileRegion(fileHandle: fileHandle)
+            var buf = try self.fileIO.read(fileRegion: fileRegion,
+                                           allocator: allocator,
+                                           eventLoop: eventLoop).wait()
+            XCTAssertEqual("0", buf.readString(length: buf.readableBytes) ?? "bad")
```
Comment (Lukasa) [inline Tests/NIOTests/NonBlockingFileIOTest.swift]: No need to nil-coalesce here: Swift will automatically promote `"0"` to `Optional<String>.some("0")`.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

File: Sources/NIO/NonBlockingFileIO.swift (hunk 1 above)
Comment (sveamarcus) [reply Sources/NIO/NonBlockingFileIO.swift]: Thanks, changes pushed

File: Sources/NIO/NonBlockingFileIO.swift [hunk 5]
```diff
@@ -359,6 +394,29 @@ public struct NonBlockingFileIO {
...
+                      toOffset: Int64,
+                      buffer: ByteBuffer,
+                      eventLoop: EventLoop) -> EventLoopFuture<()> {
+        return self.write0(fileHandle: fileHandle, toOffset: toOffset, buffer: buffer, eventLoop: eventLoop)
+    }
+
+    private func write0(fileHandle: NIOFileHandle,
+                      toOffset: Int64?,
+                      buffer: ByteBuffer,
+                      eventLoop: EventLoop) -> EventLoopFuture<()> {
```
Comment (Lukasa) [inline Sources/NIO/NonBlockingFileIO.swift]: Nit: indentation here is a bit busted.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by sveamarcus** [comment]
Sounds good :)

File: Sources/NIO/NonBlockingFileIO.swift [hunk 6]
```diff
@@ -348,6 +348,41 @@ public struct NonBlockingFileIO {
             return buf
         }
     }
+    
+    /// Writes new file size to `fileHandle` in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
```
Comment (weissi) [inline Sources/NIO/NonBlockingFileIO.swift]:
```suggestion
    /// Changes the file size of `fileHandle` to `size`.
    ///
    /// If `size` is smaller than the current file size, the remaining bytes will be truncated and are lost. If `size`
    /// is larger than the current file size, the gap will be filled with zero bytes.
```

File: Sources/NIO/NonBlockingFileIO.swift [hunk 7]
```diff
@@ -348,6 +348,41 @@ public struct NonBlockingFileIO {
...
+                               size: Int64,
+                               eventLoop: EventLoop) -> EventLoopFuture<()> {
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            try fileHandle.withUnsafeFileDescriptor { descriptor -> Void in
+                try Posix.ftruncate(descriptor: descriptor, size: off_t(size))
+            }
+        }
+    }
+
+    /// Reads the file size from `fileHandle` in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
```
Comment (weissi) [inline Sources/NIO/NonBlockingFileIO.swift]:
```suggestion
    /// Returns the length of the file associated with `fileHandle`.
```

File: Sources/NIO/NonBlockingFileIO.swift [hunk 8]
```diff
@@ -359,6 +394,29 @@ public struct NonBlockingFileIO {
     public func write(fileHandle: NIOFileHandle,
                       buffer: ByteBuffer,
                       eventLoop: EventLoop) -> EventLoopFuture<()> {
+        return self.write0(fileHandle: fileHandle, toOffset: nil, buffer: buffer, eventLoop: eventLoop)
+    }
+
+    /// Write `buffer` starting from `toOffset` to `fileHandle` in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
+    /// P
```
Comment (weissi) [inline Sources/NIO/NonBlockingFileIO.swift]: stray `P` character

File: Sources/NIO/NonBlockingFileIO.swift [hunk 9]
```diff
@@ -367,13 +425,21 @@ public struct NonBlockingFileIO {
...
-                    let res = try fileHandle.withUnsafeFileDescriptor { descriptor in
-                        try Posix.write(descriptor: descriptor,
-                                        pointer: ptr.baseAddress!,
-                                        size: byteCount)
+                    let res: IOResult<ssize_t> = try fileHandle.withUnsafeFileDescriptor { descriptor in
+                        if let toOffset = toOffset {
+                            return try Posix.pwrite(descriptor: descriptor,
+                                                    pointer: ptr.baseAddress!,
+                                                    size: byteCount,
+                                                    offset: off_t(toOffset))
```
Comment (weissi) [inline Sources/NIO/NonBlockingFileIO.swift]: this isn't correct. If `pwrite` only writes parts of the bytes, we need to increase `offset` for the subsequent calls. So you need to remember how many bytes have been written previously and add that to `toOffset`.

Comment (sveamarcus) [reply Sources/NIO/NonBlockingFileIO.swift]: Thanks, Weissi, that was certainly problematic. Any suitable way to unit test this?

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Comment by sveamarcus** [comment]
Thanks guys. Let's do R/W locking 😱

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Pull request merged**